src/portable/espressif/esp32s2/dcd_esp32s2.c

```diff
@@ -0,0 +1,750 @@
+    uint16_t max_size;
+    bool short_packet;
+} xfer_ctl_t;
+
+static const char *TAG = "TUSB:DCD";
+static intr_handle_t usb_ih;
+static volatile TU_ATTR_ALIGNED(4) uint32_t _setup_packet[6];
+static volatile uint8_t s_setup_phase = 0; /*  00 - got setup,
+                                               01 - got done setup,
+                                               02 - setup cmd sent*/
+
+#define XFER_CTL_BASE(_ep, _dir) &xfer_status[_ep][_dir]
+static xfer_ctl_t xfer_status[EP_MAX][2];
+
+static inline void readyfor1setup_pkg(int ep_num)
+{
+    USB0.out_ep_reg[ep_num].doeptsiz |= (1 << USB_SUPCNT0_S); // doeptsiz 29:30 will decremented on every setup received
+}
+
+// Setup the control endpoint 0.
+static void bus_reset(void)
+{
+
+    for (int ep_num = 0; ep_num < USB_OUT_EP_NUM; ep_num++) {
+        USB0.out_ep_reg[ep_num].doepctl |= USB_DO_SNAK0_M; // DOEPCTL0_SNAK
+    }
+
+    USB0.dcfg &= ~USB_DEVADDR_M; // reset address
+
+    // "USB Data FIFOs" section in reference manual
+    // Peripheral FIFO architecture
+    //
+    // --------------- 320 or 1024 ( 1280 or 4096 bytes )
+    // | IN FIFO MAX |
+    // ---------------
+    // |    ...      |
+    // --------------- y + x + 16 + GRXFSIZ
+    // | IN FIFO 2   |
+    // --------------- x + 16 + GRXFSIZ
+    // | IN FIFO 1   |
+    // --------------- 16 + GRXFSIZ
+    // | IN FIFO 0   |
+    // --------------- GRXFSIZ
+    // | OUT FIFO    |
+    // | ( Shared )  |
+    // --------------- 0
+    //
+    // According to "FIFO RAM allocation" section in RM, FIFO RAM are allocated as follows (each word 32-bits):
+    // - Each EP IN needs at least max packet size, 16 words is sufficient for EP0 IN
+    //
+    // - All EP OUT shared a unique OUT FIFO which uses
+    //   * 10 locations in hardware for setup packets + setup control words (up to 3 setup packets).
+    //   * 2 locations for OUT endpoint control words.
+    //   * 16 for largest packet size of 64 bytes. ( TODO Highspeed is 512 bytes)
+    //   * 1 location for global NAK (not required/used here).
+    //   * It is recommended to allocate 2 times the largest packet size, therefore
+    //   Recommended value = 10 + 1 + 2 x (16+2) = 47 --> Let's make it 52
+    USB0.grstctl |= 0x10 << USB_TXFNUM_S; // fifo 0x10,
+    USB0.grstctl |= USB_TXFFLSH_M;        // Flush fifo
+    USB0.grxfsiz = 52;
+
+    USB0.gintmsk = USB_MODEMISMSK_M |
+                   USB_SOFMSK_M |
+                   USB_RXFLVIMSK_M |
+                   USB_ERLYSUSPMSK_M |
+                   USB_USBSUSPMSK_M |
+                   USB_USBRSTMSK_M |
+                   USB_ENUMDONEMSK_M |
+                   USB_IEPINTMSK_M |
+                   USB_OEPINTMSK_M |
+                   USB_RESETDETMSK_M |
+                   USB_DISCONNINTMSK_M;
+
+    USB0.daintmsk |= USB_OUTEPMSK0_M | USB_INEPMSK0_M;
+    USB0.doepmsk |= USB_SETUPMSK_M | USB_XFERCOMPLMSK;
+    USB0.diepmsk |= USB_TIMEOUTMSK_M | USB_DI_XFERCOMPLMSK_M /*| USB_INTKNTXFEMPMSK_M*/;
```

Comment:
@cr1901 while troubleshoting esp32s2, I notice that this **Received IN token when FIFO is empty** (USB_INTKNTXFEMPMSK_M) mask is not enabled at the global endpoint IN. I tried to enable it and it breaks the whole driver, same thing happens with stm32 (USB_OTG_DIEPMSK_ITTXFEMSK). 
https://github.com/hathach/tinyusb/blob/master/src/portable/st/synopsys/dcd_synopsys.c#L111
May be it got triggered when data is not ready or something 

However, I am curious how we got the interrupt when it is not enabled as current code !!! According to STM32 Interrupt tree, the mask is needed !!! Maybe I misread the datasheet due to long hour of debugging, If you still remember the detail, please help to explain it. If it is indeed needed we can enable the mask and fixed the issue with it (data not ready or endpoint not enabled I guess).

![image](https://user-images.githubusercontent.com/249515/78545922-ca849b80-7826-11ea-86f6-8aa6a92d4c0c.png)

Reply:
@hathach Interesting, I agree it should be fixed. Indeed, I never enabled this interrupt on STM32. I don't have ESP32 hardware to test.

Are you saying on ESP32, the "Received IN token when FIFO is empty" interrupt triggers anyway _without_ the mask being enabled (whereas since I don't explicitly enable on STM32, the interrupt never fires)? And when you _do_ enable the mask on STM32, the driver _also_ breaks?

Reply:
@cr1901 yeah, when I enabled it on STM32 it cannot pass enumerataion, it failed somewhere. I didn't look at the detail. I think I should ask you first before troubleshoot it when esp32s2 is more reliable.  Though I think on both STM32 and ESP, the interrupt is fired even without this mask is enabled.

Reply:
>Though I think on both STM32 and ESP, the interrupt is fired even without this mask is enabled.

Of course it does... :P. Why would documentation match reality ;)?

Worth opening an issue? I can check sometime after April 15th on H7 board. I still have not replaced my fried F4 board yet.

Reply:
I open issue for it https://github.com/hathach/tinyusb/issues/331 . Feel free to check it whenever you have time, since we have absolute no issue with current code  :D. H7 is probably the same as F4, only faster :D.